pkg/metadata/store/metadata_store.go

```diff
@@ -0,0 +1,231 @@
+type UpdateFunc func([]byte) ([]byte, error)
+
+// MetadataStore is the interface for storing metadata.
+// TODO(random-liu): Initialize the metadata store with a type, and replace
+// []byte with interface{}, so as to avoid extra marshal/unmarshal on the
+// user side.
+type MetadataStore interface {
+	// Create the metadata containing the passed in data with the
+	// specified id.
+	// Note:
+	// * Create MUST return error if the id already exists.
+	// * The id and data MUST be added in one transaction to the store.
+	Create(string, []byte) error
+	// Get the data by id.
+	// Note that Get MUST return nil without error if the id
+	// doesn't exist.
+	Get(string) ([]byte, error)
+	// Update the data by id. Note that the update MUST be applied in
+	// one transaction.
+	Update(string, UpdateFunc) error
+	// List returns entire array of data from the store.
+	List() ([][]byte, error)
+	// Delete the data by id.
+	// Note:
+	// * Delete should be idempotent, it MUST not return error if the id
+	// doesn't exist or has been removed.
+	// * The id and data MUST be deleted in one transaction.
+	Delete(string) error
+}
+
+// TODO(random-liu) Add checkpoint. When checkpoint is enabled, it should cache data
+// in memory and checkpoint metadata into files during update. Metadata should serve
+// from memory, but any modification should be checkpointed, so that memory could be
+// recovered after restart. It should be possible to disable the checkpoint for testing.
+// Note that checkpoint update may fail, so the recovery logic should tolerant that.
+
+// An atomic file operator is needed.
+
+// metadata is unit stored in metadataStore.
+type metadata struct {
+	sync.RWMutex
+	data []byte
+}
+
+// newMetadata creates a new metadata.
+func newMetadata(data []byte) (*metadata, error) {
+	return &metadata{data: data}, nil
+	// TODO(random-liu): Create the data on disk atomically.
+}
+
+// get a snapshot of the metadata.
+func (m *metadata) get() []byte {
+	m.RLock()
+	defer m.RUnlock()
+	return m.data
+}
+
+// update the value.
+func (m *metadata) update(u UpdateFunc) error {
+	m.Lock()
+	defer m.Unlock()
+	newData, err := u(m.data)
+	if err != nil {
+		return err
+	}
+	// Replace with newData, user holding the old data will not
+	// be affected.
+	// TODO(random-liu) *Update* existing data on disk atomically,
+	// return error if checkpoint failed.
+	m.data = newData
+	return nil
+}
+
+// delete deletes the data on disk atomically.
+func (m *metadata) delete() error {
+	// TODO(random-liu): Hold write lock, rename the data on the disk.
+	return nil
+}
+
+// cleanup cleans up all temporary files left-over.
+func (m *metadata) cleanup() error {
+	// TODO(random-liu): Hold write lock, Cleanup temporary files generated
+	// in atomic file operations. The write lock makes sure there is no on-going
+	// update, so any temporary files could be removed.
+	return nil
+}
+
+// metadataStore is the implementation of MetadataStore
```

Comment:
`metadataStore implements MetadataStore` or `metadataStore is an implementation of MetadataStore`